File: extensions/shared/src/main/java/app/morphe/extension/youtube/patches/AppRefreshRatePatch.java
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches/pull/2695
 *
 * See the included NOTICE file for GPLv3 Section 7 terms that apply to this code.
 */

package app.morphe.extension.youtube.patches;

import android.app.Activity;

import app.morphe.extension.shared.patches.BaseAppRefreshRatePatch;
import app.morphe.extension.youtube.shared.PlayerType;
import app.morphe.extension.youtube.shared.VideoState;
import kotlin.Unit;

@SuppressWarnings("unused")
public final class AppRefreshRatePatch {

    /**
     * Injection point.
     */
    public static void initialize(Activity activity) {
        if (!BaseAppRefreshRatePatch.isPatchEnabled()) {
            return;
        }

        VideoState.getOnChange().addObserver((VideoState state) -> {
            updatePlayerIsActive(PlayerType.getCurrent(), state);
            return Unit.INSTANCE;
        });

        PlayerType.getOnChange().addObserver((PlayerType type) -> {
            updatePlayerIsActive(type, VideoState.getCurrent());
            return Unit.INSTANCE;
        });
    }

    private static void updatePlayerIsActive(PlayerType type, VideoState state) {
        final boolean isPlaying = state == VideoState.PLAYING;
        BaseAppRefreshRatePatch.setVideoPlayerIsActive(
                isPlaying && type == PlayerType.WATCH_WHILE_MAXIMIZED,
                isPlaying && type == PlayerType.WATCH_WHILE_FULLSCREEN
        );
    }
}

File: patches/src/main/kotlin/app/morphe/patches/youtube/misc/refreshrate/AppRefreshRatePatch.kt
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches/pull/2695
 *
 * See the included NOTICE file for GPLv3 Section 7 terms that apply to this code.
 */

package app.morphe.patches.youtube.misc.refreshrate

import app.morphe.patcher.extensions.InstructionExtensions.addInstruction
import app.morphe.patches.shared.misc.refreshrate.baseAppRefreshRatePatch
import app.morphe.patches.youtube.shared.YouTubeActivityOnCreateFingerprint
import app.morphe.patches.youtube.utils.compatibility.Constants.COMPATIBILITY_YOUTUBE
import app.morphe.patches.youtube.utils.extension.sharedExtensionPatch
import app.morphe.patches.youtube.utils.patch.PatchList.APP_REFRESH_RATE
import app.morphe.patches.youtube.utils.settings.ResourceUtils.addPreference
import app.morphe.patches.youtube.utils.settings.settingsPatch

private const val EXTENSION_CLASS = "Lapp/morphe/extension/youtube/patches/AppRefreshRatePatch;"

@Suppress("unused")
val appRefreshRatePatch = baseAppRefreshRatePatch(
    APP_REFRESH_RATE.title,
    APP_REFRESH_RATE.summary,
    block = {
        dependsOn(
            sharedExtensionPatch,
            settingsPatch
        )

        compatibleWith(COMPATIBILITY_YOUTUBE)
    },
    executeBlock = {
        addPreference(
            arrayOf(
                "SETTINGS: APP_REFRESH_RATE"
            ),
            APP_REFRESH_RATE
        )

        YouTubeActivityOnCreateFingerprint.method.addInstruction(
            0,
            "invoke-static/range { p0 .. p0 }, $EXTENSION_CLASS->" +
                    "initialize(Landroid/app/Activity;)V",
        )
    }
)
